src/Microsoft.AspNet.Razor/CodeGenerators/CSharpTagHelperCodeRenderer.cs

```diff
@@ -251,69 +226,134 @@ protected virtual string GenerateUniqueId()
-                        .WriteStringLiteral(attributeName)
-                        .WriteParameterSeparator()
-                        .Write(valueAccessor)
-                        .WriteEndMethodInvocation();
+                    // We capture the value location of each bound attribute to ensure when multiple TagHelpers request
+                    // the same attribute we don't execute the attributes value more than once.
+                    string valueLocation = null;
+
+                    foreach (var associatedDescriptor in associatedDescriptors)
+                    {
+                        // There should only be a single TagHelperAttributeDescriptor associated with the attribute.
+                        // TagHelpers cannot have multiple attribute descriptors pointing to the same property unless
+                        // they're indexed.
+                        var associatedAttributeDescriptor = associatedDescriptor.Attributes.First(
+                            attributeDescriptor => attributeDescriptor.IsNameMatch(attribute.Key));
+                        var tagHelperVariableName = GetVariableName(associatedDescriptor);
+
+                        valueLocation = RenderBoundAttribute(
+                            tagHelperVariableName,
+                            valueLocation,
+                            attribute,
+                            associatedAttributeDescriptor);
+                    }
                 }
                 else
                 {
-                    // The attribute value has already been recorded, lets retrieve it from the stored value
-                    // accessors.
-                    _writer
-                        .WriteStartAssignment(valueAccessor)
-                        .Write(htmlAttributeValues[attributeName])
-                        .WriteLine(";");
+                    RenderUnboundAttribute(attribute);
+                }
+            }
+        }
+
+        private string RenderBoundAttribute(
+            string tagHelperVariableName,
+            string valueLocation,
+            KeyValuePair<string, Chunk> attribute,
+            TagHelperAttributeDescriptor attributeDescriptor)
+        {
+            var attributeName = attribute.Key;
+            var attributeValueChunk = attribute.Value;
+            if (attributeValueChunk == null)
+            {
+                // Minimized attributes are not valid for bound attributes. TagHelperBlockRewriter has already
+                // logged an error if it was a bound attribute; so we can skip.
+                return null;
+            }
+
+            // We need to capture the tag helper's property value accessor so we can retrieve it later (if we need to).
+            var valueAccessor = string.Format(
+                CultureInfo.InvariantCulture,
+                "{0}.{1}",
+                tagHelperVariableName,
+                attributeDescriptor.PropertyName);
```

Comment:
Why bother when `valueLocation` is non-`null`? This should be within the `if` block below.